### ISSUE TITLE
Allow the use of mirrors.ubuntu.com

### ISSUE THREAD

**Issue by solsticedhiver** [xolox/python-apt-mirror-updater#6]
Instead of using https://launchpad.net/ubuntu/+archivemirrors, one could use instead http://mirrors.ubuntu.com.

It features a list of mirrors per country, or gives a list of mirrors given your IP, if using http://mirrors.ubuntu.com/mirrors.txt

These lists are way shorter than the whole list at launchpad.net, and then are faster to test for the best mirror for example, given that a mirror in your country is most of the time closer to you than the others.

The only downside is that it is using http instead of https.

**Comment by xolox** [comment]
It's funny that you're bringing this up. The following bullet points explain why (until the vertical bar, after that I'l respond more directly to your suggestion):

- Starting from the first release of `apt-mirror-updater` on 2016-03-10 mirror discovery for Ubuntu did in fact use [mirrors.ubuntu.com/mirrors.txt](http://mirrors.ubuntu.com/mirrors.txt), this "magical URL" provided the starting point for me to launch the `apt-mirror-updater` project!

- This worked fine until sometime around 2017-06-08 (about fifteen months later) when I discovered that [mirrors.ubuntu.com/mirrors.txt](http://mirrors.ubuntu.com/mirrors.txt) had stopped working as expected, now serving a text file with a single URL that wasn't useful to me at all. The commit 54d86d5fd74b6040f99444402bfcbffe9d850247 is about this, see also the related [failure of build 240920748](https://travis-ci.org/xolox/python-apt-mirror-updater/builds/240920748), here's the most relevant bit from the logs:

```
2017-06-08 19:24:04 ... Fetching http://mirrors.ubuntu.com/mirrors.txt ..
2017-06-08 19:24:05 ... Took 0.23 seconds to fetch http://mirrors.ubuntu.com/mirrors.txt.
2017-06-08 19:24:05 ... Discovered 1 Ubuntu mirror in 0.23 seconds.
```

- In response to this unexpected change in behavior I started searching for "official documentation" on the behavior of [mirrors.ubuntu.com/mirrors.txt](http://mirrors.ubuntu.com/mirrors.txt) and I found nothing. It was completely unclear to me if this was a transient error or if the service had effectively been taken down and was never coming back...

- Because of the sudden failure and lack of transparency about why it was failing I decided that I shouldn't have picked an undocumented (possibly internal and/or deprecated) service of the Ubuntu project to base my work on, so I instead developed code to scrape the official overview of mirrors on Launchpad (the officially documented source of Ubuntu mirrors) in 7175d2ccc56624a9b5d536826c19027cb5942b16. This was released in the form of `apt-mirror-updater` 1.0 in 7d22a6d1aa3cdad68a321ecb45886a6ff1267f16 and has worked fine ever since.

----

Your feedback suggests that [mirrors.ubuntu.com/mirrors.txt](http://mirrors.ubuntu.com/mirrors.txt) is working properly again and you would prefer using that over scraping Launchpad. However I prefer for `apt-mirror-updater` to be reliable "do what I mean" software and [mirrors.ubuntu.com/mirrors.txt](http://mirrors.ubuntu.com/mirrors.txt) has proven to be unreliable in the past. So what can we do? Well, let's see:

- I think Launchpad is more reliable.
- You think [mirrors.ubuntu.com/mirrors.txt](http://mirrors.ubuntu.com/mirrors.txt) is faster.

I think the best thing I can do is to change the Ubuntu mirror discovery as follows:

1. Discover Ubuntu mirrors on Launchpad.
2. Try to discover mirrors using [mirrors.ubuntu.com/mirrors.txt](http://mirrors.ubuntu.com/mirrors.txt) and iff successful, narrow down the list produced in step 1 based on the URLs reported in step 2.
3. Rank the discovered / narrowed down mirrors and pick the best one.

Does that make sense? The reason why I'm considering adding this kind of complexity is because it has bothered me in the past that Ubuntu mirror discovery was slow. This would help.

**Comment by xolox** [comment]
This is implemented in `apt-mirror-updater` version 5.2 which was just released to GitHub and PyPI, I hope this helps. If you find any issues with the new version feel free to reopen this issue or open a new one. Thanks for the feedback!

**Comment by solsticedhiver** [comment]
wow that was fast. I did not get the time to comment.

In fact, my idea was that you provide that as an option under an argument of your script, like --use-ubuntu-mirror or something like that. It wouldn't have change at all your code logic. just another mirror to choose from and fetch the mirrors.

But ok. Thanks anyway.
